### PR TITLE
[SPARK-25117][R] Add EXEPT ALL and INTERSECT ALL support in R

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
[SPARK-21274](https://issues.apache.org/jira/browse/SPARK-21274) added support for EXCEPT ALL and INTERSECT ALL. This PR adds the support in R.

## How was this patch tested?
Added test in test_sparkSQL.R

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -2848,6 +2848,35 @@ setMethod("intersect",
...
+#' @name intersectAll
+#' @examples
+#'\dontrun{
+#' sparkR.session()
+#' df1 <- read.json(path)
+#' df2 <- read.json(path2)
+#' intersectAllDF <- intersectAll(df1, df2)
+#' }
+#' @rdname intersectAll
+#' @note intersectAll since 2.4
```
Comment (felixcheungu) [inline R/pkg/R/DataFrame.R]: please put `2.4.0`

Comment (dilipbiswal) [reply R/pkg/R/DataFrame.R]: @felixcheungu Ok.

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 2]
```diff
@@ -2482,6 +2482,32 @@ test_that("union(), unionByName(), rbind(), except(), and intersect() on a DataF
...
 })
 
+test_that("intersectAll() and exceptAll()", {
+  df1 <- createDataFrame(
+    list(list("a", 1),
+      list("a", 1),
+      list("a", 1),
+      list("a", 1),
+      list("b", 3),
+      list("c", 4)),
```
Comment (HyukjinKwon) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: nit:

```r
list(list("a", 1), list("a", 1), list("a", 1),
     list("a", 1), list("b", 3), list("c", 4)),
```

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 3]
```diff
@@ -2482,6 +2482,32 @@ test_that("union(), unionByName(), rbind(), except(), and intersect() on a DataF
...
+      schema = c("a", "b"))
+  df2 <- createDataFrame(
+    list(list("a", 1), list("a", 1), list("b", 3)),
+    schema = c("a", "b"))
```
Comment (HyukjinKwon) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: nit:

```r
df2 <- createDataFrame(list(list("a", 1), list("a", 1), list("b", 3)), schema = c("a", "b"))
```

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 4]
```diff
@@ -2482,6 +2482,32 @@ test_that("union(), unionByName(), rbind(), except(), and intersect() on a DataF
...
+  intersect_all_expected <- data.frame("a" = c("a", "a", "b"), "b" = c(1, 1, 3),
+                                       stringsAsFactors = FALSE)
+  except_all_expected <- data.frame("a" = c("a", "a", "c"), "b" = c(1, 1, 4),
+                                    stringsAsFactors = FALSE)
+  intersect_all_df <- arrange(intersectAll(df1, df2), df1$a)
```
Comment (HyukjinKwon) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: Strictly, the naming rule is `intersectAllDf` or `intersect.all.df` (see https://github.com/apache/spark/pull/17590#issuecomment-293732796)

**Comment by HyukjinKwon** [comment]
Seems fine.

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/R/DataFrame.R [hunk 5]
```diff
@@ -2876,6 +2905,37 @@ setMethod("except",
...
+#' @rdname exceptAll
+#' @name exceptAll
+#' @examples
+#'\dontrun{
+#' sparkR.session()
+#' df1 <- read.json(path)
+#' df2 <- read.json(path2)
+#' exceptAllDF <- exceptAll(df1, df2)
+#' }
+#' @rdname exceptAll
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: this is a bug in `except` there should only be one `@rdname` for each

File: R/pkg/R/DataFrame.R [hunk 6]
```diff
@@ -2848,6 +2848,35 @@ setMethod("intersect",
...
+#' @rdname intersectAll
+#' @name intersectAll
+#' @examples
+#'\dontrun{
+#' sparkR.session()
+#' df1 <- read.json(path)
+#' df2 <- read.json(path2)
+#' intersectAllDF <- intersectAll(df1, df2)
+#' }
+#' @rdname intersectAll
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: ditto here

File: R/pkg/R/DataFrame.R [hunk 7]
```diff
@@ -2848,6 +2848,35 @@ setMethod("intersect",
...
+#' @note intersectAll since 2.4.0
+setMethod("intersectAll",
+          signature(x = "SparkDataFrame", y = "SparkDataFrame"),
+          function(x, y) {
+            intersected <- callJMethod(x@sdf, "intersectAll", y@sdf)
+            dataFrame(intersected)
+          })
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: add extra empty line after code

File: R/pkg/R/DataFrame.R [hunk 8]
```diff
@@ -2876,6 +2905,37 @@ setMethod("except",
...
+#' }
+#' @rdname exceptAll
+#' @note exceptAll since 2.4.0
+setMethod("exceptAll",
+          signature(x = "SparkDataFrame", y = "SparkDataFrame"),
+          function(x, y) {
+            excepted <- callJMethod(x@sdf, "exceptAll", y@sdf)
+            dataFrame(excepted)
+          })
+
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: nit: remove one of the two empty lines

File: R/pkg/R/DataFrame.R (hunk 5 above)
Comment (dilipbiswal) [reply R/pkg/R/DataFrame.R]: @felixcheung Thanks .. Did you want the original function `except` fixed at part of this ?

File: R/pkg/R/DataFrame.R (hunk 7 above)
Comment (dilipbiswal) [reply R/pkg/R/DataFrame.R]: @felixcheung OK.

File: R/pkg/R/DataFrame.R (hunk 8 above)
Comment (dilipbiswal) [reply R/pkg/R/DataFrame.R]: @felixcheung Sure.

**Comment by dilipbiswal** [comment]
@felixcheung I have incorporated the comments.

**Review by felixcheung — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

**Comment by felixcheung** [comment]
merged to master

**Pull request closed without merging**